File: src/main/js/formulas/math/SUM.ts
import { isNotNull } from "../../utils/Other";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Mappers } from "../../utils/Mappers";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class SUM extends AbstractFormula {
  static SELF: SUM = new SUM();
  NAME = FormulaName.SUM;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    return values
      .map(this.lookup)
      .map(Mappers.toFlatStream)
      .reduce(Reducers.join)
      .filter(isNotNull)
      .map(Converters.toNumber)
      .reduce(Reducers.sum, 0);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/SUM.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;
import io.protobase.f7.utils.Mappers;

import java.util.Arrays;
import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Sum values together.
 */
public class SUM extends AbstractFormula {
  public static FormulaName NAME = FormulaName.SUM;
  public static SUM SELF = new SUM();

  public SUM() {
    super();
  }

  public SUM(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return Arrays.stream(values)
        .map(this.lookup)
        .flatMap(Mappers::toFlatStream)
        .filter(Objects::nonNull)
        .map(Converters::toDouble)
        .reduce(Double::sum)
        .orElse(0.0);
  }
}
